### PR TITLE
Make `AbstractGraph.nodes` and `.edges` properties, add `.edge_colors`

### PR DESCRIPTION
**Pull request by femtobit**
As discussed in #750.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@
 ### Breaking Changes
 
 * The `graph.Grid` class has been removed. {ref}`graph.Grid` will now return an instance of {ref}`graph.Lattice` supporting the same API but with new functionalities related to spatial symmetries. The `color_edges` optional keyword argument has been removed without deprecation. [#724](https://github.com/netket/netket/pull/724)
+* The `.nodes` and `.edges` methods of `graph.AbstractGraph` are now properties. For `.edges`, this now matches NetKet v2. Furthermore, the `AbstractGraph.edge_colors` property has been added (note that, in contrast to NetKet v2, it only contains the colors now without duplicating the information of `.edges`).
```
Comment (femtobit) [inline CHANGELOG.md:20]:
```suggestion
* The `.nodes` and `.edges` methods of `graph.AbstractGraph` are now properties. For `.edges`, this now matches NetKet v2. Furthermore, the `AbstractGraph.edge_colors` property has been added (note that, in contrast to NetKet v2, it only contains the colors now without duplicating the information of `.edges`). [#756](https://github.com/netket/netket/pull/756)
```

**Comment by PhilipVinc** [comment]
My only issue with that is: do we want to properly deprecate the old usage?

If so, we should do something rather ugly like wrapping the list into something that behaves like a list but is also callable (so that the sy nta x `graph.edges()` still works and gives a deprecvation warning...

```
@struct.dataclass
class callablelist(List):
    __list
   
    def __getitem__(self, i):
        return self.__list[i]

   def __add__(self, other):
       if isinstance(other, callablelist):
              other = other.__list
       return callablelist(self.__list, other)

   def __call__(self):
       warn_deprecation(...)
       return __list
```

**Comment by femtobit** [comment]
> My only issue with that is: do we want to properly deprecate the old usage?

Do we really want to go all that length to stay compatible to _both_ the NetKet v2 and previous NetKet v3 betas?

I have thought about a solution like yours, but not only is it pretty complex just for sparing people the effort of replacing `.edges()` with `.edges` right away, it is also somewhat brittle given that the types still change.

**Comment by femtobit** [comment]
Note that what you are proposing is somewhat akin to what `networkx` is actually doing: There, the edges property returns a `MultiEdgeDataView` it so that is callable (and accepts arguments such as `data="color"`, similar to what we had). So in a sense we'd be coming full circle to what originally motivated us to make `edges` a method.

**Comment by PhilipVinc** [comment]
I see your point.

The problem is that breaking code that was working fine before is a bit annoying. Especially since we changed several other things and they are all properly deprecated..

**Comment by femtobit** [comment]
Okay, maybe we should take a step back: Why did we want to change this back to properties? While it makes some sense in a vacuum, its also not a problem to keep the methods and if compatibility not just with v2 but also previous betas is important here, we maybe probably should not change it a second time now.

What originally prompted this discussion was just the observation that the `color` argument is a bit weird (though it is well-defined and can be looked up in the documentation, so that was not a strong downside and we could just document it better). No one even complained about it being different from v2 until now.

So if we actually want to change something, maybe we just slightly improve on the documentation for `color` and be done with it? Or, maybe better, split `color` into two separate arguments for filtering and returning the colors.

(And since we're at it, if necessary, I could also make the property-version kind-of work with ther "callable list" approach to provide a deprecation warning in v3.0.)

**Comment by PhilipVinc** [comment]
I keep forgetting that it was a property in netket v2.

My point is simple: the vast majority of people using netket, are they on netket v2 or on one of the many intermediate never released netket3 beta releases?

I think most people by now are in one of the many unreleased versions, with only a minority still on v2.
though it;s impossible to know...

So, if most people were still using v2, i'd agree with you and simply revert the change.

But since most people have spent a few months on those intermediate releaess, I'd prefer to give them some warning.

**Comment by femtobit** [comment]
@PhilipVinc It does sound like we agree. I can live with the current status of the v3.0 beta (that is, without this PR). The only open question is whether we want to a) change something about `color` and b) whether we want to provide some hacky (and of course temporary) backwards-compatibility with v2. Personally, I do not find either of these points crucial. (The change v2 -> v3 is essentially adding parentheses. If people are annoyed by this, they would probably be put-off more by all the changes related to the machines...)

**Comment by femtobit** [comment]
Superseded by #766.

**Pull request closed without merging**